src/contribute/runner-guide.md

```diff
@@ -0,0 +1,1103 @@
+instantiated with a `PipelineOptions` object, you access your interface by
+`options.as(MyRunnerOptions.class)`.
+
+To make these options available on the command line, you register your options
+with a `PipelineOptionsRegistrar`. It is easy if you use `@AutoService`:
+
+{:.no-toggle}
+```java
+@AutoService(PipelineOptionsRegistrar.class)
+public static class MyOptionsRegistrar implements PipelineOptionsRegistrar {
+  @Override
+  public Iterable<Class<? extends PipelineOptions>> getPipelineOptions() {
+    return ImmutableList.<Class<? extends PipelineOptions>>of(MyRunnerOptions.class);
+  }
+}
+```
+
+#### Registering your runner with SDKs for command line use
+
+To make your runner available on the command line, you register your options
+with a `PipelineRunnerRegistrar`. It is easy if you use `@AutoService`:
+
+{:.no-toggle}
+```java
+@AutoService(PipelineRunnerRegistrar.class)
+public static class MyRunnerRegistrar implements PipelineRunnerRegistrar {
+  @Override
+  public Iterable<Class<? extends PipelineRunner>> getPipelineRunners() {
+    return ImmutableList.<Class<? extends PipelineRunner>>of(MyRunner.class);
+  }
+}
+```
+ 
+### Integrating with the Python SDK
+
+In the Python SDK the registration of the code is not automatic. So there are
+few things to keep in mind when creating a new runner. 
+
+Any dependencies on packages for the new runner should be options so create a
+new target in `extra_requires` in `setup.py` that is needed for the new runner.
+
+All runner code should go in it's own package in `apache_beam/runners` directory.
+
+Register the new runner in the `create_runner` function of `runner.py` so that the
+partial name is matched with the correct class to be used.
+
+## Writing an SDK-independent runner
+
+There are two aspects to making your runner SDK-independent, able to run
+pipelines written in other languages: The Fn API and the Runner API.
+
+### The Fn API
+
+_Design documents:_
+
+ - _[https://s.apache.org/beam-fn-api](https://s.apache.org/beam-fn-api)_
+ - _[https://s.apache.org/beam-fn-api-processing-a-bundle](https://s.apache.org/beam-fn-api-processing-a-bundle)_
+ - _[https://s.apache.org/beam-fn-api-send-and-receive-data](https://s.apache.org/beam-fn-api-send-and-receive-data)_
+
+To run a user's pipeline, you need to be able to invoke their UDFs.  The Fn API
+is an RPC interface for the standard UDFs of Beam, implemented using protocol
+buffers over gRPC.
+
+The Fn API includes:
+
+ - APIs for registering a subgraph of UDFs
+ - APIs for streaming elements of a bundle
+ - Shared data formats (key-value pairs, timestamps, iterables, etc)
+
+You are fully welcome to _also_ use the SDK for your language for utility code,
+or provide optimized implementations of bundle processing for same-language
+UDFs.
+
+### The Runner API
+
+The Runner API is an SDK-independent schema for a pipeline along with RPC
+interfaces (TBD) for launching a pipeline and checking the status of a job.
```

Comment:
The "RPC" bit has not been implemented. I rephrased a little bit. I could also just drop that part or put it elsewhere.